road/scatter.py

```diff
@@ -0,0 +1,236 @@
+
+
+def compute_and_save_errors(
+        est_df,
+        obs_df,
+        chosen_timeperiod,
+        combined_df_cols,
+        classification_col):
+    # Compute errors
+    times = ['Daily', 'AM', 'MD', 'PM', 'EV', 'EA']
+    error = {col: est_df[col] - obs_df[col] for col in times}
+    error_df = pd.DataFrame(error)
+    error_squared_df = error_df.pow(2)
+    error_percent_df = error_df.divide(obs_df[times]).fillna(0)
+    error_percent_df = error_percent_df.apply(lambda x: x * 100, axis=1)
+
+    # Combine dataframes
+    combined_df = pd.concat(
+        [est_df, obs_df[times], error_df, error_squared_df, error_percent_df], axis=1)
+    Select_time_period_df = combined_df.filter(like=chosen_timeperiod)
+    specific_columns_combined_df = combined_df[combined_df_cols]
+    Select_time_period_loc_df = pd.concat(
+        [specific_columns_combined_df, Select_time_period_df], axis=1)
+
+    # Set column names
+    calculation_cols = [
+        'Estimated Volume',
+        'Observed Volume',
+        'Errors',
+        'Squared Errors',
+        'Percent Errors']
+    Select_time_period_loc_df.columns = combined_df_cols + calculation_cols
+
+    # Drop duplicates
+    Select_time_period_loc_df = Select_time_period_loc_df.drop_duplicates(
+        subset=['A', 'B'], keep='first')
+
+    # Save to CSV
+    file_name = f"{chosen_timeperiod}_Scatter_data.csv"
+    Select_time_period_loc_df.to_csv(file_name, index=False)
+
+    # Get classification column types
+    classification_col_types = Select_time_period_loc_df[classification_col].unique(
+    )
+
+    return Select_time_period_loc_df, classification_col_types, file_name
+
+
+def sanitize_filename(filename):
+    return filename.replace('/', '_').replace('\\', '_')
+
+
+def generate_vega_lite_json_est(
+        csv_file,
+        classification_col,
+        classification_col_types,
+        x_field,
+        y_field,
+        fields,
+        nominal_fields,
+        name,
+        include_all_data=False):
+    df = pd.read_csv(csv_file)
+    if not include_all_data:
+        df = df[df[classification_col] == classification_col_types]
+    max_value = max(df[x_field].max(), df[y_field].max())
+    model = LinearRegression().fit(df[[x_field]], df[y_field])
+    m = model.coef_[0]
+    b = model.intercept_
+
+    regression_df = pd.DataFrame({
+        x_field: [df[x_field].min(), df[x_field].max()],
+        y_field: [m * df[x_field].min() + b, m * df[x_field].max() + b]
+    })
+    diagonal_df = pd.DataFrame({
+        x_field: [df[x_field].min(), df[x_field].max()],
+        y_field: [df[x_field].min(), df[x_field].max()]  # Diagonal line: y = x
+    })
+    equation_text = f"y = {m:.2f}x + {b:.2f}"
+    max_value = max(df[x_field].max(), df[y_field].max())
+    tooltip_config = [{"field": field,
+                       "type": "nominal" if field in nominal_fields else "quantitative",
+                       "title": field} for field in fields]
+

```

Comment:
Let's separate out the calculations and the output (i.e. above and below this line) into 2 separate functions? Thoughts?

Reply:
yes we can, I just wonder if it is really needed, cause  only the est/obs scatter plot need a regression line

Reply:
I see, I guess it comes down to preference. a 100-line function just seems super long. But I defer to your judgement

Reply:
I see, separating this part does make the code clear and easy to check.  Btw I just remember if the Simwrapper's regression functionality works, we probably don't need this part.  I think the problem is the data complexity exceeded Simwrapper's calculation capacity. (The data was not that large though)